crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -1,40 +1,79 @@
+use crate::mapping::{
+    converter::*, error::*, operations::Operations, size_threshold::SizeThreshold,
+};
+use agent_interface::{
+    topic::{RequestTopic, ResponseTopic},
+    Auth, DownloadInfo, Jsonify, OperationStatus, RestartOperationRequest,
+    RestartOperationResponse, SoftwareListRequest, SoftwareListResponse, SoftwareUpdateResponse,
+};
+use async_trait::async_trait;
+use c8y_smartrest::{
+    alarm,
+    error::SmartRestDeserializerError,
+    smartrest_deserializer::{SmartRestRestartRequest, SmartRestUpdateSoftware},
+    smartrest_serializer::{
+        CumulocitySupportedOperations, SmartRestGetPendingOperations, SmartRestSerializer,
+        SmartRestSetOperationToExecuting, SmartRestSetOperationToFailed,
+        SmartRestSetOperationToSuccessful, SmartRestSetSupportedLogType,
+        SmartRestSetSupportedOperations,
+    },
+};
 use c8y_translator::json;
-use mqtt_channel::{Message, Topic};
-use std::collections::hash_map::Entry;
-use std::collections::{HashMap, HashSet};
-use std::fs::File;
-use std::io::Read;
-use std::path::Path;
+use mqtt_channel::{Message, Topic, TopicFilter};
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    fs::File,
+    io::Read,
+    path::Path,
+    process::Stdio,
+};
 use thin_edge_json::alarm::ThinEdgeAlarm;
-use tracing::info;
+use tracing::{debug, info, log::error};
+
+use super::{
+    error::SMCumulocityMapperError,
+    fragments::{C8yAgentFragment, C8yDeviceDataFragment},
+    http_proxy::C8YHttpProxy,
+    json_c8y::C8yUpdateSoftwareListResponse,
+    mapper::CumulocityMapper,
+    topic::{C8yTopic, MapperSubscribeTopic},
+};
 
 const C8Y_CLOUD: &str = "c8y";
 const INVENTORY_FRAGMENTS_FILE_LOCATION: &str = "/etc/tedge/device/inventory.json";
-const INVENTORY_MANAGED_OBJECTS_TOPIC: &str = "c8y/inventory/managedObjects/update";
 const SUPPORTED_OPERATIONS_DIRECTORY: &str = "/etc/tedge/operations";
+const INVENTORY_MANAGED_OBJECTS_TOPIC: &str = "c8y/inventory/managedObjects/update";
 const SMARTREST_PUBLISH_TOPIC: &str = "c8y/s/us";
 const TEDGE_ALARMS_TOPIC: &str = "tedge/alarms/";
 const INTERNAL_ALARMS_TOPIC: &str = "c8y-internal/alarms/";
 
-pub struct CumulocityConverter {
+#[derive(Debug)]
+pub struct CumulocityConverter<Proxy>
+where
+    Proxy: C8YHttpProxy,
+{
     pub(crate) size_threshold: SizeThreshold,
     children: HashSet<String>,
     pub(crate) mapper_config: MapperConfig,
     device_name: String,
     device_type: String,
-
     alarm_converter: AlarmConverter,
+    operations: Operations,
+    http_proxy: Proxy,
 }
 
-impl CumulocityConverter {
-    pub fn new(size_threshold: SizeThreshold, device_name: String, device_type: String) -> Self {
-        let topics = vec![
+impl<Proxy> CumulocityConverter<Proxy>
+where
+    Proxy: C8YHttpProxy,
+{
+    pub fn new(
+        size_threshold: SizeThreshold,
+        device_name: String,
+        device_type: String,
+        operations: Operations,
+        http_proxy: Proxy,
```

Comment:
Doesn't this injection of http_proxy into the converter corrupt the original vision for these `Converter`s to stay stateless and protocol agnostic? Shouldn't the `Proxy`, which is a comms thing just like the mqtt channels, stay in the `c8y/mapper` module rather than here? And have the mapper do the switching between MQTT or HTTP based on the message type returned by the converter (an mqtt message or http message)?

May be this is outside the scope of this PR and this protocol switching logic needs to be implemented separately. Even the `Converter` needs be enhanced so that it just doesn't return a vector of mqtt messages but can have http messages as well in them.

Reply:
This is true and therefore as soon as #844 is merged this can go away.